---
title: "Using peakPeekeR"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Using-peakPeekeR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(peakPeekeR)
library(rtracklayer)
library(Rsamtools)
library(GenomicAlignments)
```

## Testing BAM to BigWig and subsetting.

```{r}
gr <- GRanges(seqnames = "chr12",
              ranges = IRanges(start = c(6522378), end = c(6769097)))

bammy <- BamFile("~/scratch/tester/A549_H3K27AC.ENCFF872MQN.sorted.bam", index = "~/scratch/tester/A549_H3K27AC.ENCFF872MQN.sorted.bam.bai")
inny <- BamFile("~/scratch/tester/A549_INPUT.ENCFF928UTV.sorted.bam", index = "~/scratch/tester/A549_INPUT.ENCFF928UTV.sorted.bam.bai")

params <- ScanBamParam(which = gr, what = scanBamWhat())

tmp <- tempdir()

start.time <- Sys.time()

trt <- readGAlignments(bammy, param = params)
ctrl <- readGAlignments(bammy, param = params)

trt_cov <- coverage(trt)
ctrl_cov <- coverage(ctrl)

export.bw(trt_cov, con = paste0(tmp, "/trt.bw"))
export.bw(ctrl_cov, con = paste0(tmp, "/ctrl.bw"))

end.time <- Sys.time()
end.time - start.time
```

Test of plotting peak call beds and bw files in plotgardener
```{r}
library("plotgardener")
library("plotgardenerData")
library("org.Hs.eg.db")
library("TxDb.Hsapiens.UCSC.hg19.knownGene")

data("GM12878_ChIP_CTCF_signal")

#peak <- ("ATAC-peak_test.bed")
#peak_df <- read.table(peak)
## ATAC label
bw <- ("/rgs01/project_space/bakergrp/baker_data/bakergrp/internal/3D_consortium/Unaltered_CellLines/ATACseq/processed/nfcore_runs/OMNI_040122/nfcore_hg19/bwa/mergedLibrary/bigwig/SJHGG031038_C21-LTC77_P12_R1.mLb.clN.bigWig")


## Create a plotgardener page
pageCreate(width = 5, height = 2.25, default.units = "inches")

## Text section label
plotText(label = "A", fontsize = 12,
         x = 0.25, y = 0.25, just = "left", default.units = "inches")

## Set genomic and dimension parameters in a `params` object
params_c <- pgParams(chrom = "chr21", chromstart = 28150000, chromend = 29150000, 
                     assembly = "hg19",
                     x = 0.5, width = 4, default.units = "inches")
## Set signal track data ranges
ctcf_range <- pgParams(range = c(0, 77),
                       assembly = "hg19")
hk_range <- pgParams(range = c(0, 32.6),
                     assembly = "hg19")

bw_range <- pgParams(range = c(0, 0.5),
                     assembly = "hg19")

peak_range <- pgParams(assembly = "hg19")
              
## Plot CTCF signal
ctcf_gm <- plotSignal(data = GM12878_ChIP_CTCF_signal, params = c(params_c, ctcf_range),
                      fill = "#253494", linecolor = "#253494",
                      y = 0.25, height = 0.6)
## CTCF label
plotText(label = "CTCF", fontcolor = "#253494", fontsize = 8,
         x = 1, y = 0.25, just = c("left","top"), default.units = "inches")

  
## Plot ATAC big wig
ATAC_bw <- plotSignal(data = bw, params = c(params_c, bw_range),
                      fill = "#253494", linecolor = "#253494",
                      y = 1, height = 0.6)

## ATAC label
plotText(label = "ATAC big wig", fontcolor = "#253494", fontsize = 8,
         x = 1, y = 1, just = c("left","top"), default.units = "inches")

## Plot ATAC peak calls
# ATAC_peak <- plotRanges(data = peak_df, params = c(params_c, peak_range),
#                          fill = "#253494", linecolor = "#253494",
#                          y = 1.75, height = 0.4)
## ATAC peak labels label
plotText(label = "ATAC peaks", fontcolor = "#253494", fontsize = 8,
         x = 1, y = 1.75, just = c("left","top"), default.units = "inches")

## Plot genes
genes_gm <- plotGenes(params = params_c, stroke = 1, fontsize = 6,
                      strandLabels = FALSE,
                      y = 2.25, height = 0.4)
## Annotate genome label
annoGenomeLabel(plot = genes_gm, params = params_c, 
                scale = "Kb", fontsize = 7,
                y = 3)

## Hide page guides
pageGuideHide()

```

```{r}
library(ggbio)
library(BSgenome.Hsapiens.UCSC.hg38)
bs <- BSgenome.Hsapiens.UCSC.hg38
gr <- GRanges(seqnames = "chr12",
              ranges = IRanges(start = c(6522378), end = c(6769097)))

autoplot(bammy, which = gr, bsgenome = bs)
```

